### ISSUE TITLE
message: Maximum call stack size exceeded

### ISSUE THREAD

**Issue by pine** [lukeed/taskr#141]
Hi. I try to use Fly v0.9.1, but it doesn't work in part of environments Node v4 and npm v2.
I develop [fly-ava](https://github.com/pine613/fly-ava) plugins, and also use it, so I've found errors as the following.

```
message: Maximum call stack size exceeded
stack: 
  - RangeError: Maximum call stack size exceeded
  -     at Function.isArray (native)
  -     at reduce (~/build/pine613/fly-ava/node_modules/fly/dist/utils/bind.js 37:13)
  -     at reduce (~/build/pine613/fly-ava/node_modules/fly/dist/utils/bind.js 43:14)
  -     at reduce (~/build/pine613/fly-ava/node_modules/fly/dist/utils/bind.js 43:14)
  -     at reduce (~/build/pine613/fly-ava/node_modules/fly/dist/utils/bind.js 43:14)
  -     at reduce (~/build/pine613/fly-ava/node_modules/fly/dist/utils/bind.js 43:14)
  -     at reduce (~/build/pine613/fly-ava/node_modules/fly/dist/utils/bind.js 43:14)
  -     at reduce (~/build/pine613/fly-ava/node_modules/fly/dist/utils/bind.js 43:14)
  -     at reduce (~/build/pine613/fly-ava/node_modules/fly/dist/utils/bind.js 43:14)
  -     at reduce (~/build/pine613/fly-ava/node_modules/fly/dist/utils/bind.js 43:14)
```

https://travis-ci.org/pine613/fly-ava/jobs/110563430

I think that `reduce` util function has bugs.
https://github.com/bucaran/fly/blob/master/src/utils/bind.js#L21-L33

This code runs forever.

```
reduce([]);
```

Thank you.

**Comment by lukeed** [comment]
Apologies for this.

I'm already aware of this. It's an easy fix, but still unusable afterwards because of Babel compilation on the fly...

The next version of Fly (coming soon) drops Babel completely and so these silly issues won't be a problem anymore.

If you're going to continue using Node 4.x, I'd suggest rolling back to Fly v0.8.8.
Otherwise, you can upgrade to Node 5.x, and proceed using Fly v0.9

**Comment by ralyodio** [comment]
how will you support async/await without babel?

**Comment by ghost** [comment]
@chovy Good question, we don't know yet :sweat_smile: 

Personally, I would vote against OTTB support for `async/await`, but that does not mean we will drop it so chances are we will end up keeping it.

**Comment by pine** [comment]
> The next version of Fly (coming soon) drops Babel completely and so these silly issues won't be a problem anymore.

Wow!
I am looking forward to it!

Thank you.

**Comment by lukeed** [comment]
:) it's already done and being thoroughly beta-tested now. So far it's all working smoothly  (and 60% quicker)!

**Comment by IvanSanchez** [comment]
I had this exact same problem when using Fly 0.9, but it's gone in the 1.0.0-beta4. Maybe @pine613 can try again and see if it is still reproducible?

**Comment by jrop** [comment]
I'm using fly 1.0.1, and I'm still getting this issue (please re-open this issue!).  If I change `lib/utils/logging.js:trace(e)` to look like:

``` js
function trace(e) {
    console.log(e.stack)
}
```

Then it works.  The problem seems to be an issue with `prettyjson` trying to render the whole object, and it errors-out with a stack overflow.

**Comment by lukeed** [comment]
What is your Node version?

**Comment by jrop** [comment]
I am using Node v5.5.0

**Comment by lukeed** [comment]
OK, and what triggers the stack trace? Pls post relevant flyfile task

**Comment by jrop** [comment]
I am using the `fly-webpack` plugin, with this task:

``` js
$['public:js'] = function * () {
    yield this.source('src/public/js/**/*.js')
        .webpack(require('./webpack.config.js'))
}
```

And my webpack.config.js:

``` js
module.exports = {
    devtool: 'inline-source-map',
    module: {
        loaders: [ {
            test: /\.jsx?$/,
            exclude: /node_modules/,
            loader: 'babel',
        } ]
    },
    output: {
        filename: 'build/public/js/[name].js'
    }
}
```

What happens is Webpack complains that it can't find a file (I forgot a `..` in my path), and when fly tries to print out a stack trace using `prettyjson`, it causes the stack overflow

**Comment by lukeed** [comment]
Can you post the full trace output? With the errors

**Comment by jrop** [comment]
If I `console.log(e)`, I get:

```
ModuleNotFoundError: Module not found: Error: Cannot resolve 'file' or 'directory' ../react in src/public/js
    at node_modules/webpack/lib/Compilation.js:229:38
    at onDoneResolving (node_modules/webpack/lib/NormalModuleFactory.js:29:20)
    at node_modules/webpack/lib/NormalModuleFactory.js:85:20
    at node_modules/async/lib/async.js:726:13
    at node_modules/async/lib/async.js:52:16
    at done (node_modules/async/lib/async.js:241:17)
    at node_modules/async/lib/async.js:44:16
    at node_modules/async/lib/async.js:723:17
    at node_modules/async/lib/async.js:167:37
    at node_modules/enhanced-resolve/lib/UnsafeCachePlugin.js:24:19
```

and If I `console.log(require('util').inspect(e))`:

```
{ [ModuleNotFoundError: Module not found: Error: Cannot resolve 'file' or 'directory' ../react in src/public/js]
  name: 'ModuleNotFoundError',
  message: 'Module not found: Error: Cannot resolve \'file\' or \'directory\' ../react in src/public/js',
  details: 'resolve file\n  src/public/react doesn\'t exist\n  src/public/react.webpack.js doesn\'t exist\n  src/public/react.web.js doesn\'t exist\n  src/public/react.js doesn\'t exist\n  src/public/react.json doesn\'t exist\nresolve directory\n  src/public/react doesn\'t exist (directory default file)\n  src/public/react/package.json doesn\'t exist (directory description file)',
  missing: 
   [ 'src/public/react',
     'src/public/react.webpack.js',
     'src/public/react.web.js',
     'src/public/react.js',
     'src/public/react.json' ],
  module: 
   DependenciesBlock {
     dependencies: 
      [ [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object] ],
     blocks: [],
     variables: [],
     context: 'src/public/js',
     reasons: [ [Object] ],
     debugId: 1001,
     lastId: -1,
     id: 1,
     index: 1,
     index2: 215,
     chunks: [ [Object] ],
     warnings: [],
     dependenciesWarnings: [],
     errors: [],
     dependenciesErrors: [ [Circular] ],
     request: 'node_modules/babel-loader/index.js!src/public/js/app.js',
     userRequest: 'src/public/js/app.js',
     rawRequest: 'src/public/js/app.js',
     parser: 
      Parser {
        _plugins: [Object],
        options: undefined,
        scope: undefined,
        state: undefined,
        _currentPluginApply: undefined },
     resource: 'src/public/js/app.js',
     loaders: [ 'node_modules/babel-loader/index.js' ],
     fileDependencies: [ 'src/public/js/app.js' ],
     contextDependencies: [],
     error: null,
     _source: 
      SourceMapSource {
        _value: '\'use strict\';\n\nvar _react = require(\'react\');\n\nvar _react2 = _interopRequireDefault(_react);\n\nvar _reactDom = require(\'react-dom\');\n\nvar _reactDom2 = _interopRequireDefault(_reactDom);\n\nvar _reactRouter = require(\'react-router\');\n\nvar _react3 = require(\'../react\');\n\nfunction _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }\n\nwindow.addEventListener(\'load\', function () {\n\t_reactDom2.default.render(_react2.default.createElement(\n\t\t_reactRouter.Router,\n\t\t{ history: _reactRouter.hashHistory },\n\t\t_react2.default.createElement(\n\t\t\t_reactRouter.Route,\n\t\t\t{ path: \'/\', component: App },\n\t\t\t_react2.default.createElement(_reactRouter.Route, { path: \'/dar\', component: _react3.Dar }),\n\t\t\t_react2.default.createElement(_reactRouter.Route, { path: \'/dm\', component: _react3.Dm })\n\t\t)\n\t), document.getElementById(\'react\'));\n});',
        _name: 'node_modules/babel-loader/index.js!src/public/js/app.js',
        _sourceMap: [Object],
        _originalSource: undefined,
        _innerSourceMap: undefined },
     meta: {},
     assets: {},
     built: true,
     _cachedSource: { source: [Object], hash: 'b356653224bdbf6c43ba63108ab9a8bd' },
     issuer: 'multi app',
     optional: false,
     useSourceMap: true,
     building: undefined,
     buildTimestamp: 1458156216660,
     cacheable: true },
  error: 
   { [Error: Cannot resolve 'file' or 'directory' ../react in src/public/js]
     details: 'resolve file\n  src/public/react doesn\'t exist\n  src/public/react.webpack.js doesn\'t exist\n  src/public/react.web.js doesn\'t exist\n  src/public/react.js doesn\'t exist\n  src/public/react.json doesn\'t exist\nresolve directory\n  src/public/react doesn\'t exist (directory default file)\n  src/public/react/package.json doesn\'t exist (directory description file)',
     missing: 
      [ 'src/public/react',
        'src/public/react.webpack.js',
        'src/public/react.web.js',
        'src/public/react.js',
        'src/public/react.json' ] },
  dependencies: 
   [ CommonJsRequireDependency {
       module: null,
       request: '../react',
       userRequest: '../react',
       range: [Object],
       loc: [Object],
       optional: false } ],
  origin: 
   DependenciesBlock {
     dependencies: 
      [ [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object] ],
     blocks: [],
     variables: [],
     context: 'src/public/js',
     reasons: [ [Object] ],
     debugId: 1001,
     lastId: -1,
     id: 1,
     index: 1,
     index2: 215,
     chunks: [ [Object] ],
     warnings: [],
     dependenciesWarnings: [],
     errors: [],
     dependenciesErrors: [ [Circular] ],
     request: 'node_modules/babel-loader/index.js!src/public/js/app.js',
     userRequest: 'src/public/js/app.js',
     rawRequest: 'src/public/js/app.js',
     parser: 
      Parser {
        _plugins: [Object],
        options: undefined,
        scope: undefined,
        state: undefined,
        _currentPluginApply: undefined },
     resource: 'src/public/js/app.js',
     loaders: [ 'node_modules/babel-loader/index.js' ],
     fileDependencies: [ 'src/public/js/app.js' ],
     contextDependencies: [],
     error: null,
     _source: 
      SourceMapSource {
        _value: '\'use strict\';\n\nvar _react = require(\'react\');\n\nvar _react2 = _interopRequireDefault(_react);\n\nvar _reactDom = require(\'react-dom\');\n\nvar _reactDom2 = _interopRequireDefault(_reactDom);\n\nvar _reactRouter = require(\'react-router\');\n\nvar _react3 = require(\'../react\');\n\nfunction _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }\n\nwindow.addEventListener(\'load\', function () {\n\t_reactDom2.default.render(_react2.default.createElement(\n\t\t_reactRouter.Router,\n\t\t{ history: _reactRouter.hashHistory },\n\t\t_react2.default.createElement(\n\t\t\t_reactRouter.Route,\n\t\t\t{ path: \'/\', component: App },\n\t\t\t_react2.default.createElement(_reactRouter.Route, { path: \'/dar\', component: _react3.Dar }),\n\t\t\t_react2.default.createElement(_reactRouter.Route, { path: \'/dm\', component: _react3.Dm })\n\t\t)\n\t), document.getElementById(\'react\'));\n});',
        _name: 'node_modules/babel-loader/index.js!src/public/js/app.js',
        _sourceMap: [Object],
        _originalSource: undefined,
        _innerSourceMap: undefined },
     meta: {},
     assets: {},
     built: true,
     _cachedSource: { source: [Object], hash: 'b356653224bdbf6c43ba63108ab9a8bd' },
     issuer: 'multi app',
     optional: false,
     useSourceMap: true,
     building: undefined,
     buildTimestamp: 1458156216660,
     cacheable: true } }
```

**Comment by lukeed** [comment]
Can you post the output with the original `trace` method? I want to see the stack that causes the `Maximum call stack size exceeded`

**Comment by jrop** [comment]
The original `trace` method:

``` js
function trace(e) {
    var msg = require('util').inspect(e)
        pretty.render(e)
        .replace(/(\sFunction|\sObject)\./g, clor.blue('$1') + '.')
        .replace(/\((~?\/.*)\)/g, '(' + clor.gray('$1') + ')')
        .replace(/:([0-9]*):([0-9]*)/g, ' ' + clor.yellow('$1') + ':' + clor.yellow('$2'))
        .replace(new RegExp(process.env.HOME, 'g'), '~')
    console.log(msg)
}
```

**Comment by lukeed** [comment]
Actually, this is the `trace` method:

``` js
function trace(e) {
    var msg = pretty.render(e)
        .replace(/(\sFunction|\sObject)\./g, clor.blue('$1') + '.')
        .replace(/\((~?\/.*)\)/g, '(' + clor.gray('$1') + ')')
        .replace(/:([0-9]*):([0-9]*)/g, ' ' + clor.yellow('$1') + ':' + clor.yellow('$2'))
        .replace(new RegExp(process.env.HOME, 'g'), '~')
    console.error(msg)
}
```

Please paste the unmodified terminal output when using :point_up_2: the original trace method

**Comment by jrop** [comment]
Yup, I just noticed, that, my bad...I've been messing with this some :P 

The stack when using your original trace method is:

```
message:   Maximum call stack size exceeded
undefined: true
stack: 
  - RangeError: Maximum call stack size exceeded
  -     at Object.stylize (~/.nvm/v5.5.0/lib/node_modules/fly/node_modules/colors/lib/colors.js 55:33)
  -     at String.green (~/.nvm/v5.5.0/lib/node_modules/fly/node_modules/colors/lib/extendStringPrototype.js 58:21)
  -     at ~/.nvm/v5.5.0/lib/node_modules/fly/node_modules/prettyjson/lib/prettyjson.js 148:18
  -     at Array.forEach (native)
  -     at Object.render (~/.nvm/v5.5.0/lib/node_modules/fly/node_modules/prettyjson/lib/prettyjson.js 144:38)
  -     at ~/.nvm/v5.5.0/lib/node_modules/fly/node_modules/prettyjson/lib/prettyjson.js 131:31
  -     at Array.forEach (native)
  -     at Object.render (~/.nvm/v5.5.0/lib/node_modules/fly/node_modules/prettyjson/lib/prettyjson.js 114:12)
  -     at ~/.nvm/v5.5.0/lib/node_modules/fly/node_modules/prettyjson/lib/prettyjson.js 166:19
  -     at Array.forEach (native)
```

**Comment by lukeed** [comment]
Lol, thanks. I use Node v5.5 & Fly v1.\* daily, and I've never run into this, and I get errors all the time :innocent: 

So I'm leaning towards it being something `fly-webpack` related... but I'll give it due diligence and check it out fully.

**Comment by jrop** [comment]
Yeah, that seems entirely possible, perhaps I will go open an issue over with fly-webpack.

Thanks for being so responsive!

**Comment by lukeed** [comment]
What would be ideal is `fly-webpack` throwing an error, instead of catching with `cb(err)`.

``` js
this.emit('plugin_error', {
  plugin: 'fly-webpack',
  error: err.message // or pre-format the output text here
})
```

**Comment by lukeed** [comment]
@jrop are you still experiencing this? Did you happen to implement the Gist I sent you?

**Comment by lukeed** [comment]
@jrop @pine Anyone still experiencing this with an updated Fly?

**Comment by jrop** [comment]
@lukeed Sorry, haven't used fly in a while, so the status of this is unknown to me

**Comment by lukeed** [comment]
@jrop All good, just doing some spring cleaning

**Comment by lukeed** [comment]
Closing for inactivity & no longer relevant to `2.x`
